### ISSUE TITLE
Proposal: Regular Naming Convention for Systems, Services and Service Functions

### ISSUE THREAD

**Issue by emanuelpalm** [eclipse-arrowhead/roadmap#4]
Note that this is not the same as the naming convention established in [(Paniagua, 2019)](https://ieeexplore.ieee.org/abstract/document/8972250). This is about how to come up with names, not with how to structure and spread them via DNS-SD.

As a programmer, I have been taught that consistency in naming is a boon to productivity. If you have a systematic way to name your artifacts, you will instantly know what type it has when you see its name. I wish Eclipse Arrowhead would have the same.

I propose the following:

## System Names

__Description__: A system can be seen as something capable of fulfilling a role. As a consequence, I believe systems should always have [agent nouns](https://en.wikipedia.org/wiki/Agent_noun) as names. An agent noun is the agent or title form of a verb. In English, such nouns typically end with "-or", "-ar", "-er" or "-ist". The verb indicates what the system does when fulfilling its role. It being in its agent noun form means that you are referring to the doer rather than the action itself.
__Examples__: _Service Registrar_, _Authorizer_, Orchestrator, Gatekeeper, _On-Boarder_, and _Certificate Authorizer_.
__Exceptions__: The current Gateway system is kind of an odd bird, as it does not provide any regular services at all (if not counting the echo and management services). As it can only function in tandem with a Gatekeeper system, I don't really mind it keep being called Gateway. The word Gateway gives it an obvious connection to the Gatekeeper system, which I also like.

## Service Names

__Description__: A service represents a certain task a system is capable of performing. I believe services should have [verbal nouns](https://en.wikipedia.org/wiki/Verbal_noun), or [other nouns with derivable verbs](https://en.wikipedia.org/wiki/Deverbal_noun), as names. Agent nouns are, of course, off limits, even though they qualify as _deverbal_ nouns. The name is a noun to indicate that it _can_ be performed rather than it _being_ performed. The name being related to a verb means that it becomes more apparent what is happening when the service is invoked. Suitable endings of such verbal nouns could be "-ment", "-ing", "-y" or "-ion".
__Examples__: Service Discovery, Authorization, Orchestration, _Gatekeeping_, On-Boarding, and Authorization Management.

## Service Function Names

__Description__: A service function  represents a certain action a system is capable of performing as part of it fulfilling the task associated with one of its services. I believe service function names should be in the [imperative mood](https://en.wikipedia.org/wiki/Imperative_mood), which means that they take the form of commandments. This is customary for naming functions in most programming languages.
__Examples__: Authorize, On-Board, Register, Discover, Lookup and Qualify.
__Exceptions__: This does not mean that RESTful service implementations should have endpoints like `POST /authorize`. The SDs should have the imperative mood names, while the corresponding RESTful IDDs should stick to the REST convention and use enpoints like `POST /authorizations` (which adds another `authorization` object to an imagined folder of authorizations).

## Service Function Type Names

__Description__: A service function type names some kind of message, or structure or primitive that can be part of a message sent to or returned by a service function. Following regular programming convention, these should have names being proper nouns, optionally with preceding adjectives.
__Examples__: Authorization Request, On-Boarding Request, _Service Record_, and _Orchestration Record_.

## Words with specific Arrowhead definitions

It is typically wise to avoid using names that can become ambiguous in system contexts. Examples of such names are "Manager", "Handler" and "Data". The reason for this is that all systems manage data, and handling is synonymous with managing. What does a _Data Handler_ do? It is not very obvious from the name, as all computers handle data. How exactly does an _Event Handler_ handle event? Does it also receive them?

However, as Arrowhead is a new thing, we are free to give general words such as these very specific meanings. For example, it is already the case that the word "Management" in a service name indicates that the service is meant to allow for administration and configuration of the system in question. Why not formalize it? @jenseliasson Could you think of a similarly precise definition for the word "Handler"?

**Comment by emanuelpalm** [comment]
## These Names Need Not be Bound by ASCII

Among programmers, it is customary to stick to a subset of the ASCII character set (`[A-Za-z_][0-9A-Za-z_]*`) when naming artifacts. The reason for this is historical, as many programming languages have decided to stick to the limitations imposed by early programming languages and character sets. I want to propose, however, that the naming of systems, services, service functions and service function types are not bound by this convention, but are instead encouraged to use a form closer to regular English, which means that spaces, colons and commas may be used in these names (even I don't really see any reason why the two latter ever would be relevant).

In other words, these names would be preferred (especially in documentation):

Certificate Authorizer, Device Registrar, Event Publisher.


While these forms of the same names would be advised against (but may, of course, be used in computer code, configuration files or wherever it would be required by the programming language or context in general):

CertificateAuthorizer, DeviceRegistrar, EventPublisher.

**Comment by emanuelpalm** [comment]
@jerkerdelsing What would the process be through which a proposal like this is ratified or rejected? Would we have a ballot in a Roadmap WG meeting? The regular Bi-Weekly meeting? Who would be allowed cast a vote? Everyone present?

**Comment by jerkerdelsing** [comment]
We should think about this in relation to important standards like:
- ISO 10303
- ISO 15926
- BIM
etc.

**Comment by jerkerdelsing** [comment]
Let's discuss this at age Roadmap session of the Vaccine WS next week.

**Comment by emanuelpalm** [comment]
@jerkerdelsing Were there any discussions on this topic at the Vaccine WS or afterwards?

**Comment by jerkerdelsing** [comment]
The connection standards was mentioned again. The reson why is that this is how things, sw and services are namned in different type of industrial installations. The basis is standards like to one mentioned in my previous comment.

**Comment by emanuelpalm** [comment]
@jerkerdelsing I tried to look for applicable parts of the ISO 10303 and 16926 standards. EXPRESS (ISO 10303-11:2004) was the only part with anything that seemed to be remotely applicable. I do not have access to it, however, since it is behind a paywall. Could you direct me to where to look for the naming convention you are referring to?

**Comment by jerkerdelsing** [comment]
A number of these standards and the specific parts concerning naming of things and services have been collected in Arrowhead Tools Owncloud. https://atmospheres.research.ltu.se/owncloud/index.php/apps/files/?dir=/Arrowhead_Tools/WP3%20Arrowhead%20Framework/Task%203.1/Work/SysML_Arrowhead_profile/Profile/JD/Datamodel%20standards&fileid=1944843

**Comment by emanuelpalm** [comment]
I have now scanned through the following documents:

1. [EXPRESS 1.1](https://www.omg.org/spec/EXPRESS/1.1/PDF)
2. [UML 2.5.1](https://www.omg.org/spec/UML/2.5.1/PDF)
3. ISO/IEC 9594-8:2020 (a.k.a. ITU-T __X.509__ (10/2019))
4. ISO 10303-55:2005
5. ISO 15926-2:2003
6. ISO 16757-2:2019
7. ISO 23387:2020 (__BIM__)
8. ISO/TS 81346-3:2012

Some of these documents are concerned with the structure of identifiers, such as ISO/TS 81346-3:2012. Others with naming specific entities (as opposed to groups of related entities, as we are), such as ISO 16757-2:2019. None of these documents are concerned about word endings, whether nouns should be used, etc. In other words, I do not see how we would be diverging from any of these standards by, for example, requiring that systems be given names that are agent nouns.

That being said, the names of the structure elements I use here is something we could use EXPRESS for, in particular. By structure element, I refer to the idea of __systems__ containing __services__, which in turn are comprised of __functions__ that take __function types__ as input and return one such type as output. Systems and services are not something I believe to be relevant to even consider changing, the other names _function_ and _function type_ are, however, not something I perceive as being well-grounded. In EXPRESS, these are referred to as __algorithms__ and __types__. More specifically, there are two types of algorithms, __functions__ and __procedures__. They are different by only functions being able to produce a return value. There is quite a few different types, but their various forms and characteristics have no bearing on what I propose here.

**Comment by emanuelpalm** [comment]
@jerkerdelsing I changed the titles "Service Function Names" and "Service Function Type Names" to "Algorithm Names" and "Type Names" to better conform to EXPRESS terminology.

**Comment by emanuelpalm** [comment]
An alternative to calling them "algorithms" is to take on the name used by UML, "operations", which is used to name behavioral features associated with classes.

**Comment by jerkerdelsing** [comment]
A complex problem overall. We can almost always find another used approach for us to use. Using something well known is interesting but it need to serve our purpose and requirement.

**Comment by emanuelpalm** [comment]
@jerkerdelsing Could we use [SysML 2](https://github.com/Systems-Modeling/SysML-v2-Release) as our goto-standard for terminology? If we intend to SysML 2 it extensively it could be a good way to avoid having to map between concepts.

**Comment by vanDeventer** [comment]
I prefer "operation".
Going back to your first post, end up with: System name, Service name, Operation name. A system offers services that performs operations. A system consumes services that perform operations.

**Comment by emanuelpalm** [comment]
@vanDeventer I agree that "operation" is a better name than function. I will update the "Core Concepts" document (previously known as the "Eclipse Arrowhead Reference Model") to use the word.

We still have not come to any conclusion regarding the naming of the actual systems, services and operations, however.

**Comment by jenseliasson** [comment]
Hi.
I also like the term 'operation.

Jens

**Comment by jerkerdelsing** [comment]
If we think the interoperability way we should be able to handle any approach to name a service, a service function, a system and a device. The different application area standards have various background like e.g. the Express language (10303 and 15926), KernML (SysML 2). We probably need to stick to one way of typing a service, service function, system and device within Arrowhead. 

For the "service function" discussed I agree with using the type "operation" which matches SysML 1.6. I'm not sure about the SysML 2. Will ask Oystein.

**Comment by emanuelpalm** [comment]
@jerkerdelsing I have been unable to find any material in EXPRESS or SysML2 that specify what kind of grammatical spoken-language constructs should be used when naming anything. Also, I don't understand why any such names would lead to any kinds of interoperability-issues with any existing standards. We are not talking about identifiers here, but about names read and written by humans. Only identifiers matching the regex `[A-Za-Z][0-9A-Za-z_]*` can be defined in EXPRESS. SysML2 allows for any UTF-8 characters to be used, even spaces. This as identifiers can be written surrounded by simple quotes (`'`). Is this what you are having in mind? If so, we could just demand that my above guidelines be followed but that the regex `[A-Za-Z](_? [0-9A-Za-z])*` is satisfied by the name. When names are written for humans only, underscores are substituted for plain spaces. This should guarantee compatibility with pretty much every modeling format and programming language there is.

SysML2 seems to have changed the name from "Operation" to "Function" (see [Kernel Modeling Language version 1.0 (ad/2021-11-02), Section 7.4.7](https://raw.githubusercontent.com/Systems-Modeling/SysML-v2-Release/master/doc/1-Kernel_Modeling_Language.pdf#%5B%7B%22num%22%3A157%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C652%2C0%5D)).

**Comment by jerkerdelsing** [comment]
@emanuelpalm You should not find that there. But since EXPESS and KernML is used for higher level languages like ISO 10303 AP242 or AP209, ISO 15926, BIM, ... who each of them have a specification for how for example a sensor should be named. These specifications are driven by the application field and here will interoperability problems show up when things like IoT's need to talk to another IoT across an application domain boarder.

**Comment by emanuelpalm** [comment]
@jerkerdelsing Nothing is preventing us from both having Arrowhead-centric names and other names for the same entities. For example, I noted above that adhering to my convention for operation names is not compatible with the RESTful specification. This is not a problem, as long as a mapping is made between the Arrowhead names and the REST names in IDDs. Supporting multiple names for specific entities is probably something we will want in either case, since a use case may demand that a sensor, for example, both satisfies ISO 10303 AP242 and some other standard with an incompatible naming schema.
